<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
</head>

<body>
  <script>
    // value 记录成功的原因 reason 记录失败的原因
    // promise 状态只能改变一次 pending -> fullfilled penging -> rejected  不能完成态 拒绝态相互转换
    // promise 在执行过程中  任何错误都会导致promise 变成失败态
    // promise 必须有then方法 onFulfilled：成功的回调， onRejected：失败的回调
    // promise 除了链式调用还可以多次调用
    // promise 的链式调用
    // promise 除了链式调用还可以多次调用
    const PENDING = 'pending'
    const FULFILLED = 'fullfilled'
    const REJECTED = 'rejected'
    class Promise {
      value
      reason
      status = PENDING
      // onFulfilled
      // onRejected
      onFulfilledCallBacks = []
      onRejectedCallBacks = []
      constructor(executor) {
        const resolve = (value) => {
          if (this.status === PENDING) {
            this.value = value
            this.status = FULFILLED
            // this.onFulfilled(this.value)
            this.onFulfilledCallBacks.forEach(cb => cb())
          }
        }
        const reject = (reason) => {
          if (this.status === PENDING) {
            this.reason = reason
            this.status = REJECTED
            // this.onRejected(this.reason)
            this.onRejectedCallBacks.forEach(cb => cb())
          }
        }
        try {
          executor(resolve, reject)
        } catch (error) {
          reject(error)
        }
      }
      then(onFulfilled, onRejected) {
        const p2 = new Promise((resolve, reject) => {
          if (this.status === FULFILLED) {
            queueMicrotask(() => {
              try {
                const x = onFulfilled(this.value)
                resolve(x)
              } catch (error) {
                reject(error);
              }
            })
          }
          if (this.status === REJECTED) {
            queueMicrotask(() => {
              try {
                const x = onRejected(this.reason)
                resolve(x)
              } catch (error) {
                reject(error)
              }
            })
          }
          if (this.status === PENDING) {
            // this.onFulfilled = onFulfilled
            // this.onRejected = onRejected
            this.onFulfilledCallBacks.push(() => {
              queueMicrotask(() => {
                try {
                  const x = onFulfilled(this.value)
                  resolvePromise((x,p2,resolve,reject))
                  resolve(x)
                } catch (error) {
                  reject(error)
                }
              })
            })
            this.onRejectedCallBacks.push(() => {
              queueMicrotask(() => {
                try {
                  const x = onRejected(this.reason)
                  resolve(x)
                } catch (error) {
                  reject(error)
                }
              })
            })
          }
        })
        return p2
      }
    }
    // 用x定p2的状态
    // x 是then 成功或者失败回调的返回值
    // p2 then方法返回的promise
    // 
    function resolvePromise(x,p2,resolve,reject) {
      if(x===2) {
        throw new TypeError('chaining cycle detected for promise')
      }
        
    }
    const p = new Promise((resolve, reject) => {
      // resolve('成功')
      setTimeout(() => { 

        resolve('成功')
      }, 3000)
    })

    p.then(
      (res) => console.log('第一次'),
    )

    p.then(
      (res) => console.log('第二次'),
    )

  </script>
</body>

</html>